server/src/main/java/io/crate/planner/optimizer/Rule.java

```diff
@@ -33,7 +35,12 @@
 
     Pattern<T> pattern();
 
-    LogicalPlan apply(T plan, Captures captures, TableStats tableStats, TransactionContext txnCtx, NodeContext nodeCtx);
+    LogicalPlan apply(T plan,
+                      Captures captures,
+                      TableStats tableStats,
+                      TransactionContext txnCtx,
+                      NodeContext nodeCtx,
+                      Function<LogicalPlan, LogicalPlan> resolveLogicalPlan);
```

Comment:
```suggestion
    /**
     * @param resolveLogicalPlan {@code plan} Could contain proxies as source. Must use this function on the sources to resolve them to the actual nodes before they can be used
     **/
    LogicalPlan apply(T plan,
                      Captures captures,
                      TableStats tableStats,
                      TransactionContext txnCtx,
                      NodeContext nodeCtx,
                      Function<LogicalPlan, LogicalPlan> resolveLogicalPlan);
```

Or something like this? I think I don't fully understand when it should be used and when not - looking through the rules there are a couple where it's not used, while in others it is and I don't really understand the pattern/logic behind that.